ocaml/xapi/test_network_event_loop.ml

```diff
@@ -0,0 +1,165 @@
+let test_network_event_loop ~no_nbd_networks_at_start () =
+  let __context, _ = Test_event_common.event_setup_common () in
+  (* We need to set test_mode to true to ensure that the event loop will use
+     the local Xapi_events.from function instead of the Client module, which
+     would cause the test to fail. *)
+  Helpers.test_mode := true;
+
+  let localhost = Helpers.get_localhost ~__context in
+  let other_host = Test_common.make_host ~__context () in
+
+  (* We have to wait for a bit for the event loop to notice the changes, without a delay the test will fail. *)
+  let delay = 0.2 in
+  let network_event_loop_wait_after_failure_seconds = 0.4 in
+  let received_params = ref None in
+
+  (* We simulate failure of the firewall update script this way *)
+  let fail_firewall_update = ref false in
+
+  let start_event_loop = Thread.create
+      (fun () ->
+         Network_event_loop._watch_networks
+           __context
+           ~update_firewall:(fun pifs ->
+               if !fail_firewall_update then failwith "Failed to update firewall";
+               received_params := Some pifs
+             )
+           ~wait_after_failure_seconds:network_event_loop_wait_after_failure_seconds
+      )
+  in
+
+  let assert_received_params msg expected =
+    Thread.delay delay;
+    match !received_params with
+    | None -> OUnit.assert_failure ("The update_firewall function was not called: " ^ msg)
+    | Some p ->
+      Ounit_comparators.StringSet.(assert_equal (of_list expected) (of_list p))
+  in
+
+  let assert_not_called msg () =
+    Thread.delay delay;
+    OUnit.assert_equal ~msg:("update_firewall shouldn't have been called: " ^ msg) None !received_params
+  in
+
+  let network1 = Test_common.make_network ~__context ~bridge:"bridge1" () in
+
+  if no_nbd_networks_at_start then begin
+    received_params := None;
+    let _ : Thread.t = start_event_loop () in
+    assert_received_params "The event loop should always update the firewall state at startup, even if no networks are allowed for NBD" []
+  end;
+
+  (* Add a new network "network1" with a PIF connected to this host *)
```

Comment:
The comment here is wrong: the network exists already but doesn't have a PIF connected to this host.